### PR TITLE
Update pyproject.toml

### PR DESCRIPTION
**Pull request by jchen351**
### Description
Update pyproject.toml and remove outdate python version support. Adding python313.

### CODE REVIEW

**Review by tianleiwu — DISMISSED** [review]
please merge main to fix build.

**Comment by jchen351** [comment]
Black dosen't support py313.

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -43,7 +43,7 @@ jobs:
         uses: actions/setup-python@v5
         with:
           # Version range or exact version of Python to use, using SemVer's version range syntax. Reads from .python-version if unset.
-          python-version: "3.10"
+          python-version: "3.12"
```
Comment (snnn) [inline .github/workflows/lint.yml:46]: Have you tried to change it to 3.13?  Did it work?

File: pyproject.toml [hunk 2]
```diff
@@ -2,7 +2,7 @@
 line-length = 120
 # NOTE: Do not extend the exclude list. Edit .lintrunner.toml instead
 extend-exclude = "cmake|onnxruntime/core/flatbuffers/"
-target-version = ["py37", "py38", "py39", "py310", "py311"]
+target-version = ["py310","py311","py312","py313"]
```
Comment (snnn) [inline pyproject.toml:5]: If py313 doesn't work, we don't have to enable it now.

**Review by tianleiwu — APPROVED** [review]

**Comment by snnn** [comment]
Why the linter runner found so many issues?

**Comment by tianleiwu** [comment]
> Why the linter runner found so many issues?
I think it is because the target changes from py38 to py310. May be try run `lintrunner -a` to format files and add those changed files.

**Comment by justinchuby** [comment]
The target version for ruff was bumped to 3.10, which means rules that are only applicable to higher python versions are now enabled. These rules are necessary to keep our code base up-to-date. For example, for compatibility reasons we needed to import `Dict` for type annotations in 3.8, but in `3.10` we can use `dict` as a generic type. 

I would recommend doing this PR in two steps. 

1. Leave `target-version` for `ruff` as 3.8 and commit the rest
2. Update `target-version` for `ruff` to 3.10, run `lintrunner -a --all-files --take RUFF` to apply all changes in a separate PR.

File: pyproject.toml [hunk 3]
```diff
@@ -2,7 +2,7 @@
 line-length = 120
 # NOTE: Do not extend the exclude list. Edit .lintrunner.toml instead
 extend-exclude = "cmake|onnxruntime/core/flatbuffers/"
-target-version = ["py37", "py38", "py39", "py310", "py311"]
+target-version = ["py310","py311","py312"]
```
Comment (justinchuby) [inline pyproject.toml:5]:
```suggestion
target-version = ["py310", "py311", "py312"]
```

nit: formatting

**Comment by jchen351** [comment]
> Sorry let me update my previous comment:
> 
> 1. It looks like all that's updated is for the formatter and linters. In this case, I would: do the black and ruff changes and update the desired versions. Apply lintrunner changes (`lintrunner f --all-files`) and commit them here in the same PR (this PR). These are linter changes.
> 2. For package metadata changes, it looks like you need to update https://github.com/microsoft/onnxruntime/blob/62f99d8a8d4470520f9204608af47f9162c909e8/setup.py#L539-L545
Applying `lintrunner f --all-files` will modify 300+ files.

File: setup.py [hunk 4]
```diff
@@ -537,12 +537,10 @@ def finalize_options(self):
     "Topic :: Software Development :: Libraries :: Python Modules",
     "Programming Language :: Python",
     "Programming Language :: Python :: 3 :: Only",
-    "Programming Language :: Python :: 3.7",
-    "Programming Language :: Python :: 3.8",
-    "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
     "Programming Language :: Python :: 3.12",
+    "Programming Language :: Python :: 3.13",
```
Comment (jchen351) [inline setup.py:543]: @justinchuby @snnn can you verify this changes?

Comment (snnn) [reply setup.py:543]: Yes, we should add this line.

**Comment by justinchuby** [comment]
Modifying 300+ files should be fine, because they are all auto fixes

**Pull request closed without merging**